### PR TITLE
MTA-STS Check

### PR DESCRIPTION
**Pull request by vbrown608**
This PR only includes the implementation of the actual MTA-STS check, in the name of smaller PRs. Will follow up with a PR that rolls this into the CLI and API response.

### CODE REVIEW

File: checker/mta_sts.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	if err != nil {
+		return result.Failure("Couldn't find policy file: %v", err)
+	}
+	if resp.StatusCode != 200 {
+		return result.Failure("Couldn't get policy file: %s", resp.Status)
+	}
+	// RFC says to check media type.
+	// if resp.Header["Content-Type"] != "text/plain" {
+	// 	return result.Error("Media type must be text/plain")
+	// }
```
Comment (sydneyli) [inline checker/mta_sts.go]: why commented?

File: checker/mta_sts.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+func validateMTASTSMXs(policyFileMXs []string, dnsMXs map[string]HostnameResult,
+	result CheckResult) CheckResult {
+	for _, dnsMX := range dnsMXs {
+		if !dnsMX.couldConnect() {
+			// Ignore hostnames we couldn't connect to, they may be spam traps.
+			continue
+		}
+		if !containsString(policyFileMXs, dnsMX.Hostname) {
+			result.Warning("%s appears in the DNS record but not the MTA-STS policy file",
+				dnsMX.Hostname)
```
Comment (sydneyli) [inline checker/mta_sts.go]: Wildcard policies: if one of the mxs in the policy is `*.eff.org` and we successfully connected to `mail1.eff.org`, that should pass.

File: checker/mta_sts.go (hunk 1 above)
Comment (vbrown608) [reply checker/mta_sts.go]: I was a little concerned that we'd see surprising formatting in the content type header and wasn't sure if we wanted to enforce it. I read a bit more about the expected format and also made this warning. 

From the RFC:

> When fetching a policy, senders SHOULD validate that the media type
> is "text/plain" to guard against cases where web servers allow
> untrusted users to host non-text content (typically, HTML or images)
> at a user-defined path.

**Comment by vbrown608** [comment]
Yesssss coveralls

File: checker/mta_sts.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	// Media type should be text/plain, ignoring other Content-Type parms.
+	// Format: Content-Type := type "/" subtype *[";" parameter]
+	for _, contentType := range resp.Header["Content-Type"] {
+		contentType := strings.ToLower(contentType)
+		if strings.HasPrefix(contentType, "text/plain") {
+			return result.Warning("Media type must be text/plain")
+		}
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
```
Comment (Snawoot) [inline checker/mta_sts.go]: What if client submitted domain with infinite or just very big policy file?

File: checker/mta_sts.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+		} else {
+			parsed[key] = parsed[key] + " " + value
+		}
+	}
+	return parsed
+}
+
+func checkMTASTSRecord(domain string) CheckResult {
+	result := CheckResult{Name: "mta-sts-txt"}
+	records, err := net.LookupTXT(fmt.Sprintf("_mta-sts.%s", domain))
```
Comment (Snawoot) [inline checker/mta_sts.go]: I've noticed rest of checker code use timeouts. This lookup seems to be not restricted.

File: checker/mta_sts.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+func checkMTASTSPolicyFile(domain string, hostnameResults map[string]HostnameResult) CheckResult {
+	result := CheckResult{Name: "policy_file"}
+	client := &http.Client{
+		// Don't follow redirects.
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+	policyURL := fmt.Sprintf("https://mta-sts.%s/.well-known/mta-sts.txt", domain)
+	resp, err := client.Get(policyURL)
```
Comment (Snawoot) [inline checker/mta_sts.go]: This request also has no timeouts.

**Pull request merged**